pkg/scheduler/wfq.go

```diff
@@ -260,24 +263,31 @@ func (sched *WFQScheduler) queueRequest(ctx context.Context, request Request) (q
 }
 
 // adjust queue counters. Note: qRequest pointer should not be used after calling this function as it will get recycled via Pool.
-func (sched *WFQScheduler) scheduleRequest(ctx context.Context,
-	request Request,
-	qRequest *queuedRequest,
-) (allowed bool) {
+func (sched *WFQScheduler) scheduleRequest(ctx context.Context, request Request, qRequest *queuedRequest) (allowed bool) {
 	// This request has been selected to be executed next
 	waitTime, allowed := sched.manager.Take(ctx, float64(request.Tokens))
 	// check if we need to wait
 	if allowed && waitTime > 0 {
+		// check whether ctx has deadline
+		// and if deadline is less than waitTime
+		// return tokens immediately
+		if dl, o := ctx.Deadline(); o {
+			if dl.Sub(sched.clk.Now()) < waitTime {
+				allowed = false
+				go sched.manager.Return(context.TODO(), float64(request.Tokens))
+			}
+		}
+
 		if allowed {
-			timer := time.NewTimer(waitTime)
+			timer := sched.clk.NewTimer(waitTime)
 			defer timer.Stop()
 
 			select {
 			case <-ctx.Done():
 				allowed = false
 				// return the tokens
-				sched.manager.Return(ctx, float64(request.Tokens))
-			case <-timer.C:
+				go sched.manager.Return(context.TODO(), float64(request.Tokens))
+			case <-timer.Chan():
 			}
 		}
 	}
```

Comment:
In the case of `Return` we want to do not want to use the parent context because tokens are being returned in that case and we do not case about the deadline there.